src/pycram/probabilistic_costmap.py

```diff
@@ -0,0 +1,166 @@
+from enum import Enum, auto
+from functools import cached_property
+
+import numpy as np
+import tf
+from geometry_msgs.msg import Point
+from std_msgs.msg import ColorRGBA
+from tf.transformations import quaternion_from_matrix
+from random_events.interval import closed, closed_open
+from typing_extensions import Optional, Type
+from visualization_msgs.msg import Marker, MarkerArray
+
+from . import World
+from .costmaps import Costmap, OccupancyCostmap, VisibilityCostmap
+import matplotlib.colorbar
+from .datastructures.pose import Pose
+from .ros.data_types import Duration
+from .ros.publisher import create_publisher
+from .units import meter, centimeter
+
+from pint import Unit, Quantity
+from probabilistic_model.probabilistic_circuit.nx.helper import uniform_measure_of_event, fully_factorized
+from probabilistic_model.probabilistic_circuit.nx.probabilistic_circuit import ProbabilisticCircuit
+from random_events.product_algebra import Event, SimpleEvent
+from random_events.variable import Continuous
+
+
+class Filter(Enum):
+    OCCUPANCY = auto()
+    VISIBILITY = auto()
+
+
+class ProbabilisticCostmap:
+    """
+    A Costmap that uses probability distributions for representation.
+    """
+
+    x: Continuous = Continuous("x")
+    """
+    The variable for the x-axis (height) in meters.
+    """
+
+    y: Continuous = Continuous("y")
+    """
+    The variable for the y-axis (width) in meters.
+    """
+
+    costmap: Costmap
+    """
+    The legacy costmap.
+    """
+
+    origin: Pose
+    """
+    The origin of the costmap.
+    """
+
+    size: Quantity
+    """
+    The side length of the costmap. The costmap is a square.
+    """
+
+    distribution: Optional[ProbabilisticCircuit] = None
+    """
+    The distribution associated with the costmap.
+    """
+
+    def __init__(self, origin: Pose,
+                 size: Quantity = 2 * meter,
+                 max_cells = 10000,
+                 costmap_type: Type[Costmap] = OccupancyCostmap,
+                 world: Optional[World] = None):
+
+        self.world = world if world else World.current_world
+        self.origin = origin
+        self.size = size
+
+        # calculate the number of cells per axis
+        number_of_cells = int(np.sqrt(max_cells))
+        resolution = self.size.to(meter) / number_of_cells
+
+        if costmap_type == OccupancyCostmap:
+            self.costmap = OccupancyCostmap(
+                origin=self.origin,
```

Comment:
Initializing the costmap with static arguments is very inflexible. It could be a better idea to take the parameter for the costmap as kwargs